nearcore/src/state_sync.rs

```diff
@@ -368,13 +363,17 @@ fn start_dumping(
     let epoch_info = runtime.get_epoch_info(&epoch_id)?;
     let epoch_height = epoch_info.epoch_height();
     let num_shards = runtime.num_shards(&epoch_id)?;
-    let sync_hash_block = chain.get_block(&sync_hash)?;
-    if runtime.cares_about_shard(None, sync_hash_block.header().prev_hash(), shard_id, false) {
-        assert_eq!(num_shards, sync_hash_block.chunks().len() as u64);
-        let state_root = sync_hash_block.chunks()[shard_id as usize].prev_state_root();
-        let state_root_node = runtime.get_state_root_node(shard_id, &sync_hash, &state_root)?;
+    let sync_prev_header = chain.get_block_header(&sync_hash)?;
+    let sync_prev_hash = sync_prev_header.prev_hash();
+    let prev_sync_block = chain.get_block(&sync_prev_hash)?;
+    if runtime.cares_about_shard(None, prev_sync_block.header().prev_hash(), shard_id, false) {
+        assert_eq!(num_shards, prev_sync_block.chunks().len() as u64);
```

Comment:
for this whole thing, did you consider just using `Chain::get_state_response_part()`? The logic around finding the right state roots and such is already implemented there, so you could just call that function in the main loop of `state_sync_dump()` if `InProgress` just stored the shard ID, part_id and sync_hash. Unless maybe I'm missing something

Reply:
Amazing suggestion. Done.